### PR TITLE
[cudamapper] Initial implementation of IndexGeneratorGPU

### PR DESCRIPTION
**Pull request by mimaric**

### CODE REVIEW

File: cudamapper/src/index_generator_gpu.hpp [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+#pragma once
+
+#include <cstdint>
+#include "cudamapper/index_generator.hpp"
+#include "cudamapper/sequence.hpp"
+#include "cudamapper/types.hpp"
+#include "minimizer.hpp"
+
+namespace claragenomics {
```
Comment (tijyojwad) [inline cudamapper/src/index_generator_gpu.hpp]: in our other modules we've been adding another layer of namespace for the module itself
e.g.
```
namespace claragenomics {
namespace cudamapper {
....
}
}
```

File: cudamapper/src/index_generator_gpu.hpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+        /// \brief generates data structures necessary for building the actualy index
+        ///
+        /// \param query_filename filepath to reads in FASTA or FASTQ format
+        /// \param minimizer_size k - the kmer length used as a minimizer
+        /// \param window_size w - the length of the sliding window used to find minimizer
+        IndexGeneratorGPU(const std::string& query_filename, std::uint64_t minimizer_size, std::uint64_t window_size);
+
+        /// \brief return minimizer size (k)
+        /// \return minimizer size
+        std::uint64_t minimizer_size() const;
```
Comment (tijyojwad) [inline cudamapper/src/index_generator_gpu.hpp]: just curious if the range for `uint64_t` is needed for the minimizers. we've been using signed integers as a standard, an using https://github.com/clara-genomics/ClaraGenomicsAnalysis/blob/dev-v0.2.0/common/utils/include/claragenomics/utils/signed_integer_utils.hpp#L31 to check for negative values (such as users passing negative inputs). This is just to protect from overflow situations

File: cudamapper/src/index_generator_gpu.hpp (hunk 1 above)
Comment (mimaric) [reply cudamapper/src/index_generator_gpu.hpp]: Ok, we'll fix that in a separate PR.
@vellamike FYI

File: cudamapper/src/index_generator_gpu.hpp (hunk 2 above)
Comment (mimaric) [reply cudamapper/src/index_generator_gpu.hpp]: Ok, we'll fix that in a separate PR.
@vellamike FYI

**Comment by vellamike** [comment]
@mimaric I did a debug build locally and it wasn't able to complete, debug builds are working fine though. This appears to be also happening on CI. The output was as follows:

```
/home/mike/Downloads/clion-2018.3.4/bin/cmake/linux/bin/cmake --build /home/mike/dev/ClaraGenomicsAnalysis/cmake-build-debug --target cudamapper -- -j 6
[  6%] Built target logging
[ 51%] Built target zlibstatic
[ 54%] Built target index_cpu
[ 64%] Built target index_generator_cpu
[ 70%] Built target matcher
[ 74%] Building NVCC (Device) object cudamapper/CMakeFiles/index_generator_gpu.dir/src/index_generator_gpu_generated_index_generator_gpu.cu.o
/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(842): error: identifier "window_minimizers_representation_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(844): error: identifier "window_minimizers_direction_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(846): error: identifier "window_minimizers_position_in_read_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(848): error: identifier "read_id_to_minimizers_written_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(982): error: identifier "representations_compressed_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(985): error: identifier "rest_compressed_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(1000): error: identifier "window_minimizers_representation_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(1002): error: identifier "window_minimizers_position_h" is undefined

/home/mike/dev/ClaraGenomicsAnalysis/cudamapper/src/index_generator_gpu.cu(1004): error: identifier "window_minimizers_direction_h" is undefined

9 errors detected in the compilation of "/tmp/tmpxft_00006b5e_00000000-6_index_generator_gpu.cpp1.ii".
CMake Error at index_generator_gpu_generated_index_generator_gpu.cu.o.Debug.cmake:279 (message):
  Error generating file
  /home/mike/dev/ClaraGenomicsAnalysis/cmake-build-debug/cudamapper/CMakeFiles/index_generator_gpu.dir/src/./index_generator_gpu_generated_index_generator_gpu.cu.o


cudamapper/CMakeFiles/index_generator_gpu.dir/build.make:931: recipe for target 'cudamapper/CMakeFiles/index_generator_gpu.dir/src/index_generator_gpu_generated_index_generator_gpu.cu.o' failed
make[3]: *** [cudamapper/CMakeFiles/index_generator_gpu.dir/src/index_generator_gpu_generated_index_generator_gpu.cu.o] Error 1
CMakeFiles/Makefile2:999: recipe for target 'cudamapper/CMakeFiles/index_generator_gpu.dir/all' failed
make[2]: *** [cudamapper/CMakeFiles/index_generator_gpu.dir/all] Error 2
CMakeFiles/Makefile2:1203: recipe for target 'cudamapper/CMakeFiles/cudamapper.dir/rule' failed
make[1]: *** [cudamapper/CMakeFiles/cudamapper.dir/rule] Error 2
Makefile:461: recipe for target 'cudamapper' failed
make: *** [cudamapper] Error 2
```

**Comment by mimaric** [comment]
rerun tests

**Pull request merged**